homeassistant/helpers/update_coordinator.py

```diff
@@ -25,7 +25,7 @@
 
 _T = TypeVar("_T")
 _DataUpdateCoordinatorT = TypeVar(
-    "_DataUpdateCoordinatorT", bound="DataUpdateCoordinator[Any]"
+    "_DataUpdateCoordinatorT", bound="DataUpdateCoordinator"
```

Comment:
With `bound="DataUpdateCoordinator[Any]"`, VSCode reads it as `Any`

I think the root cause is that
> TypeVar bound type cannot be generic

Taking out the generic return makes VSCode happy...

Reply:
Can you provide some more details / an example what exactly didn't work?
AFAICT pyright isn't happy about it, at least not with `typeCheckingMode` set to `strict`.

<img width="836" alt="Screenshot 2022-11-08 at 19 29 11" src="https://user-images.githubusercontent.com/30130371/200646203-c8e8d707-e091-4935-920e-8df75825ab6a.png">

Reply:
When used in an inheritance context (I think it was scrape), it kept reverting to Any instead of the specified beautifulsoup generic.

I originally tried to move it down so I could use it without quotes, and that's when I had `TypeVar bound type cannot be generic`

Reply:
I thinks yours might be a false positive from pyright

Reply:
> When used in an inheritance context (I think it was scrape), it kept reverting to Any instead of the specified beautifulsoup generic.

The bound type only effects which type can be passed for the generic not what it is inferred as if none is given. That would require https://peps.python.org/pep-0696/. I.e. it's expected that for example `self.coordinator` is inferred as `Any` if a class just inherits from `CoordinatorEntity` and not from `CoordinatorEntity[DataUpdateCoordinator[dict[str, Any]]]`.

Personally, I hope that support for PEP 696 is added soon. That's why I haven't added specific types to all `CoordinatorEntity` and `DataUpdateCoordinator` yet.

https://github.com/home-assistant/core/blob/b364ef98a073214aad8deff4ff9b91e9ff041557/homeassistant/components/airvisual/__init__.py#L338-L363

Reply:
I think this is where vscode/pylance/pyright was getting confused and couldn't infer.

https://github.com/home-assistant/core/blob/6021cedb09a2b7d3434d307bc2c124c51c8f731a/homeassistant/components/scrape/sensor.py#L171

But I'll have to confirm tomorrow

Reply:
> I think this is where vscode/pylance/pyright was getting confused and couldn't infer.

Just checked it. Everything seems fine even with `[Any]`, as expected. Mypy isn't able to infer `reveal_type(self.coordinator.data)` but that's because the types for beautifulsoup are missing. Those are shipped separately. https://pypi.org/project/types-beautifulsoup4/

> But I'll have to confirm tomorrow

👍🏻